### ISSUE TITLE
Error when running patchDetection on a SpatialExperiment

### ISSUE THREAD

**Issue by michaeljpitcher** [BodenmillerGroup/imcRtools#67]
Hi,

When running patchDetection on a SpatialExperiment object, I get the following error:

```
Error: BiocParallel errors
  1 remote errors, element index: 1
  24 unevaluated and other errors
  first remote error: Problem while computing `polygon = list(.polygon_function(x = data, coords = coords, convex = convex))`.
```

The [worked tutorial](https://bioconductor.org/packages/release/bioc/vignettes/imcRtools/inst/doc/imcRtools.html#54_Patch_detection) uses a SingleCellExperiment object (not spatial) and I've noticed if I create a SingleCellExperiment from the data in my SpatialExperiment object and add the relevant spatial coordinates ('Pos_X' and 'Pos_Y') to the colData then patchDetection does work. 

Do you know why this isn't working for my SpatialExperiment object?

**Comment by nilseling** [comment]
Hi @michaeljpitcher 

thanks for reporting this! I thought I have covered the `SpatialExperiment` use case in the [unit tests](https://github.com/BodenmillerGroup/imcRtools/blob/46388d6446e22a0afc9456d7db5f4cf431be4c1e/tests/testthat/test_patchDetection.R#L162). Could you please share the output of `show(spe)`, the function call that you used to run `patchDetection` and the output of `sessionInfo()`?

**Comment by michaeljpitcher** [comment]
Sure thing:

show(spe):
```
class: SpatialExperiment 
dim: 36 276648 
metadata(0):
assays(2): counts exprs
rownames(36): CD137 CD11b ... DNA1 DNA2
rowData names(2): Metal_tag use_channel
colnames(276648): S0555_-PT3_L_3_s0_a8_1 S0555_-PT3_L_3_s0_a8_2 ... S0576_-UH16_13 II_s0_a2_9051 S0576_-UH16_13 II_s0_a2_9052
colData names(20): sample_id img_id ... aggregatedNeighbors clustered_neighbors
reducedDimNames(1): UMAP
mainExpName: NULL
altExpNames(0):
spatialData names(2) : Pos_X Pos_Y
spatialCoords names(2) : Pos_X Pos_Y
imgData names(1): sample_id
```

Function call to run patch detection:
```
spe <- patchDetection(spe,
                      patch_cells = spe$classification.1 == CELL_TYPES$BCELLS,
                      colPairName = "neighbours",
                      expand_by = 20, 
                      img_id = "img_id",
                      BPPARAM = bpp)
```

SessionInfo:
```
R version 4.1.3 (2022-03-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Monterey 12.2.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
 [1] parallel  grid      stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] sf_1.0-7                    igraph_1.3.0                stringr_1.4.0               ggraph_2.0.5               
 [5] imcRtools_1.0.2             RColorBrewer_1.1-3          harmony_0.1.0               Rcpp_1.0.8.3               
 [9] bluster_1.4.0               BiocParallel_1.28.3         patchwork_1.1.1             scran_1.22.1               
[13] scater_1.22.0               scuttle_1.4.0               circlize_0.4.14             tidyr_1.2.0                
[17] dittoSeq_1.6.0              viridis_0.6.2               viridisLite_0.4.0           ggpubr_0.4.0               
[21] ggforce_0.3.3               doParallel_1.0.17           iterators_1.0.14            foreach_1.5.2              
[25] caret_6.0-92                lattice_0.20-45             ggplot2_3.3.5               ComplexHeatmap_2.10.0      
[29] SpatialExperiment_1.4.0     dplyr_1.0.8                 cytomapper_1.7.1            SingleCellExperiment_1.16.0
[33] SummarizedExperiment_1.24.0 Biobase_2.54.0              GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        
[37] IRanges_2.28.0              S4Vectors_0.32.4            BiocGenerics_0.40.0         MatrixGenerics_1.6.0       
[41] matrixStats_0.61.0          EBImage_4.36.0             

loaded via a namespace (and not attached):
  [1] utf8_1.2.2                shinydashboard_0.7.2      R.utils_2.11.0            tidyselect_1.1.2         
  [5] htmlwidgets_1.5.4         pROC_1.18.0               DropletUtils_1.14.2       munsell_0.5.0            
  [9] ScaledMatrix_1.2.0        units_0.8-0               codetools_0.2-18          DT_0.22                  
 [13] statmod_1.4.36            future_1.24.0             withr_2.5.0               colorspace_2.0-3         
 [17] rstudioapi_0.13           ggsignif_0.6.3            listenv_0.8.0             labeling_0.4.2           
 [21] GenomeInfoDbData_1.2.7    polyclip_1.10-0           bit64_4.0.5               farver_2.1.0             
 [25] pheatmap_1.0.12           rhdf5_2.38.1              parallelly_1.31.0         vctrs_0.4.1              
 [29] generics_0.1.2            ipred_0.9-12              R6_2.5.1                  graphlayouts_0.8.0       
 [33] ggbeeswarm_0.6.0          clue_0.3-60               rsvd_1.0.5                locfit_1.5-9.5           
 [37] concaveman_1.1.0          RTriangle_1.6-0.10        bitops_1.0-7              rhdf5filters_1.6.0       
 [41] DelayedArray_0.20.0       assertthat_0.2.1          vroom_1.5.7               promises_1.2.0.1         
 [45] scales_1.2.0              nnet_7.3-17               beeswarm_0.4.0            gtable_0.3.0             
 [49] beachmat_2.10.0           globals_0.14.0            tidygraph_1.2.1           timeDate_3043.102        
 [53] rlang_1.0.2               systemfonts_1.0.4         GlobalOptions_0.1.2       splines_4.1.3            
 [57] rstatix_0.7.0             ModelMetrics_1.2.2.2      broom_0.8.0               reshape2_1.4.4           
 [61] abind_1.4-5               backports_1.4.1           httpuv_1.6.5              tools_4.1.3              
 [65] lava_1.6.10               ellipsis_0.3.2            raster_3.5-15             proxy_0.4-26             
 [69] ggridges_0.5.3            plyr_1.8.7                sparseMatrixStats_1.6.0   zlibbioc_1.40.0          
 [73] classInt_0.4-3            purrr_0.3.4               RCurl_1.98-1.6            rpart_4.1.16             
 [77] GetoptLong_1.0.5          cowplot_1.1.1             ggrepel_0.9.1             cluster_2.1.3            
 [81] magrittr_2.0.3            data.table_1.14.2         magick_2.7.3              hms_1.1.1                
 [85] mime_0.12                 fftwtools_0.9-11          xtable_1.8-4              jpeg_0.1-9               
 [89] gridExtra_2.3             shape_1.4.6               compiler_4.1.3            tibble_3.1.6             
 [93] KernSmooth_2.23-20        crayon_1.5.1              R.oo_1.24.0               htmltools_0.5.2          
 [97] tzdb_0.3.0                later_1.3.0               tiff_0.1-11               lubridate_1.8.0          
[101] DBI_1.1.2                 tweenr_1.0.2              MASS_7.3-56               readr_2.1.2              
[105] Matrix_1.4-1              car_3.0-12                cli_3.2.0                 R.methodsS3_1.8.1        
[109] metapod_1.2.0             gower_1.0.0               pkgconfig_2.0.3           sp_1.4-6                 
[113] terra_1.5-21              recipes_0.2.0             svglite_2.1.0             vipor_0.4.5              
[117] hardhat_0.2.0             dqrng_0.3.0               XVector_0.34.0            prodlim_2019.11.13       
[121] digest_0.6.29             edgeR_3.36.0              DelayedMatrixStats_1.16.0 shiny_1.7.1              
[125] rjson_0.2.21              lifecycle_1.0.1           nlme_3.1-157              Rhdf5lib_1.16.0          
[129] carData_3.0-5             BiocNeighbors_1.12.0      limma_3.50.3              fansi_1.0.3              
[133] pillar_1.7.0              fastmap_1.1.0             survival_3.3-1            glue_1.6.2               
[137] png_0.1-7                 svgPanZoom_0.3.4          bit_4.0.4                 class_7.3-20             
[141] stringi_1.7.6             HDF5Array_1.22.1          nnls_1.4                  BiocSingular_1.10.0      
[145] e1071_1.7-9               irlba_2.3.5               future.apply_1.8.1 R version 4.1.3 (2022-03-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Monterey 12.2.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
 [1] parallel  grid      stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] sf_1.0-7                    igraph_1.3.0                stringr_1.4.0               ggraph_2.0.5               
 [5] imcRtools_1.0.2             RColorBrewer_1.1-3          harmony_0.1.0               Rcpp_1.0.8.3               
 [9] bluster_1.4.0               BiocParallel_1.28.3         patchwork_1.1.1             scran_1.22.1               
[13] scater_1.22.0               scuttle_1.4.0               circlize_0.4.14             tidyr_1.2.0                
[17] dittoSeq_1.6.0              viridis_0.6.2               viridisLite_0.4.0           ggpubr_0.4.0               
[21] ggforce_0.3.3               doParallel_1.0.17           iterators_1.0.14            foreach_1.5.2              
[25] caret_6.0-92                lattice_0.20-45             ggplot2_3.3.5               ComplexHeatmap_2.10.0      
[29] SpatialExperiment_1.4.0     dplyr_1.0.8                 cytomapper_1.7.1            SingleCellExperiment_1.16.0
[33] SummarizedExperiment_1.24.0 Biobase_2.54.0              GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        
[37] IRanges_2.28.0              S4Vectors_0.32.4            BiocGenerics_0.40.0         MatrixGenerics_1.6.0       
[41] matrixStats_0.61.0          EBImage_4.36.0             

loaded via a namespace (and not attached):
  [1] utf8_1.2.2                shinydashboard_0.7.2      R.utils_2.11.0            tidyselect_1.1.2         
  [5] htmlwidgets_1.5.4         pROC_1.18.0               DropletUtils_1.14.2       munsell_0.5.0            
  [9] ScaledMatrix_1.2.0        units_0.8-0               codetools_0.2-18          DT_0.22                  
 [13] statmod_1.4.36            future_1.24.0             withr_2.5.0               colorspace_2.0-3         
 [17] rstudioapi_0.13           ggsignif_0.6.3            listenv_0.8.0             labeling_0.4.2           
 [21] GenomeInfoDbData_1.2.7    polyclip_1.10-0           bit64_4.0.5               farver_2.1.0             
 [25] pheatmap_1.0.12           rhdf5_2.38.1              parallelly_1.31.0         vctrs_0.4.1              
 [29] generics_0.1.2            ipred_0.9-12              R6_2.5.1                  graphlayouts_0.8.0       
 [33] ggbeeswarm_0.6.0          clue_0.3-60               rsvd_1.0.5                locfit_1.5-9.5           
 [37] concaveman_1.1.0          RTriangle_1.6-0.10        bitops_1.0-7              rhdf5filters_1.6.0       
 [41] DelayedArray_0.20.0       assertthat_0.2.1          vroom_1.5.7               promises_1.2.0.1         
 [45] scales_1.2.0              nnet_7.3-17               beeswarm_0.4.0            gtable_0.3.0             
 [49] beachmat_2.10.0           globals_0.14.0            tidygraph_1.2.1           timeDate_3043.102        
 [53] rlang_1.0.2               systemfonts_1.0.4         GlobalOptions_0.1.2       splines_4.1.3            
 [57] rstatix_0.7.0             ModelMetrics_1.2.2.2      broom_0.8.0               reshape2_1.4.4           
 [61] abind_1.4-5               backports_1.4.1           httpuv_1.6.5              tools_4.1.3              
 [65] lava_1.6.10               ellipsis_0.3.2            raster_3.5-15             proxy_0.4-26             
 [69] ggridges_0.5.3            plyr_1.8.7                sparseMatrixStats_1.6.0   zlibbioc_1.40.0          
 [73] classInt_0.4-3            purrr_0.3.4               RCurl_1.98-1.6            rpart_4.1.16             
 [77] GetoptLong_1.0.5          cowplot_1.1.1             ggrepel_0.9.1             cluster_2.1.3            
 [81] magrittr_2.0.3            data.table_1.14.2         magick_2.7.3              hms_1.1.1                
 [85] mime_0.12                 fftwtools_0.9-11          xtable_1.8-4              jpeg_0.1-9               
 [89] gridExtra_2.3             shape_1.4.6               compiler_4.1.3            tibble_3.1.6             
 [93] KernSmooth_2.23-20        crayon_1.5.1              R.oo_1.24.0               htmltools_0.5.2          
 [97] tzdb_0.3.0                later_1.3.0               tiff_0.1-11               lubridate_1.8.0          
[101] DBI_1.1.2                 tweenr_1.0.2              MASS_7.3-56               readr_2.1.2              
[105] Matrix_1.4-1              car_3.0-12                cli_3.2.0                 R.methodsS3_1.8.1        
[109] metapod_1.2.0             gower_1.0.0               pkgconfig_2.0.3           sp_1.4-6                 
[113] terra_1.5-21              recipes_0.2.0             svglite_2.1.0             vipor_0.4.5              
[117] hardhat_0.2.0             dqrng_0.3.0               XVector_0.34.0            prodlim_2019.11.13       
[121] digest_0.6.29             edgeR_3.36.0              DelayedMatrixStats_1.16.0 shiny_1.7.1              
[125] rjson_0.2.21              lifecycle_1.0.1           nlme_3.1-157              Rhdf5lib_1.16.0          
[129] carData_3.0-5             BiocNeighbors_1.12.0      limma_3.50.3              fansi_1.0.3              
[133] pillar_1.7.0              fastmap_1.1.0             survival_3.3-1            glue_1.6.2               
[137] png_0.1-7                 svgPanZoom_0.3.4          bit_4.0.4                 class_7.3-20             
[141] stringi_1.7.6             HDF5Array_1.22.1          nnls_1.4                  BiocSingular_1.10.0      
[145] e1071_1.7-9               irlba_2.3.5               future.apply_1.8.1 
```

**Comment by nilseling** [comment]
Thanks! In general that looks fine to me. I'm not certain about the `spe$classification.1 == CELL_TYPES$BCELLS` call. Does it produce a logical vector of length `ncol(spe)`? And is `sum(spe$classification.1 == CELL_TYPES$BCELLS)` larger than 0?

**Comment by michaeljpitcher** [comment]
Yes to both of those:
`length(spe$classification.1 == CELL_TYPES$BCELLS) == ncol(spe)`
returns TRUE, and 
`sum(spe$classification.1 == CELL_TYPES$BCELLS)`
returns 150313

Also, if it helps, here's the code I've used to convert to a SingleCellExperiment and run patchDetection successfully:
```
sc <- SingleCellExperiment(list(counts=assay(spe, 'counts'),
                                exprs=assay(spe, 'exprs')))
colData(sc) <- colData(spe)
colData(sc)$Pos_X <- spatialCoords(spe)[,"Pos_X"]
colData(sc)$Pos_Y <- spatialCoords(spe)[,"Pos_Y"]
colPair(sc, 'neighbours') <- colPair(spe, 'neighbours')

sc <- patchDetection(sc,
                      patch_cells = sc$classification.1 == CELL_TYPES$BCELLS,
                      colPairName = "neighbours",
                      expand_by = 20, 
                      img_id = "img_id",
                      BPPARAM = bpp)
```

**Comment by michaeljpitcher** [comment]
Thanks, I'll try that but in the meantime, I've noticed something else: in the unit tests, the spe is created using SpatialExperiment:::.sce_to_spe; and therefore because Pos_X and Pos_Y are in the colData(sce) they will also present in the colData of the SpatialExperiment when converted. I've tried adding the Pos_X and Pos_Y from spatialCoords(spe) to colData(spe) and the patchDetection function does work now. So is it looking for spatial coordinates in the colData regardless of the object type?

**Comment by nilseling** [comment]
Oh of course! Thanks for pointing this out, I'll look into it!

**Comment by nilseling** [comment]
Hi @michaeljpitcher 
I have now fixed this issue in #68, sorry for that. Could you please give it a try and close this issue if it works?
Thanks again for spotting this!

**Comment by michaeljpitcher** [comment]
Tried and it works fine. Thanks!
